### PR TITLE
Generic OpenAI API provider

### PR DESCRIPTION
**Pull request by bauersimon**
Merge #27 first so we can test this and refactor.

### CODE REVIEW

**Comment by Munsio** [comment]
@bauersimon rebased the PR on the latest changes

**Comment by bauersimon** [comment]
@Munsio / @ruiAzevedo19 plz take a looksie...

File: provider/openai-api/query.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import (
+	"context"
+	"fmt"
+
+	pkgerrors "github.com/pkg/errors"
+	"github.com/sashabaranov/go-openai"
+)
+
+// QueryOpenAIModel queries an OpenAI API model.
+func QueryOpenAIAPIModel(ctx context.Context, client *openai.Client, modelIdentifier string, promptText string) (response string, err error) {
```
Comment (Munsio) [inline provider/openai-api/query.go:12]: Wouldn't it be easier to just embed the openAI provider in Ollama/Openrouter?

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 2]
```diff
@@ -528,6 +529,54 @@ func TestEvaluateExecute(t *testing.T) {
...
+						var err error
+						shutdown, err = tools.OllamaStart(logger, tools.OllamaPath, tools.OllamaURL)
+						require.NoError(t, err)
+
+						require.NoError(t, tools.OllamaPull(logger, tools.OllamaPath, tools.OllamaURL, "qwen:0.5b"))
+					},
+
+					Arguments: []string{
+						"--language", "golang",
+						"--urls", "custom-ollama:" + ollamaOpenAIAPIUrl,
```
Comment (Munsio) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:559]: nitpick: would extract `custom-ollama` and `qwen:0.5b` into variables as they are quite often used and changing it anywhere could lead to unwanted sideffects.

File: provider/openai-api/openai.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+var _ provider.Query = (*Provider)(nil)
+
+// Query queries the provider with the given model name.
+func (p *Provider) Query(ctx context.Context, modelIdentifier string, promptText string) (response string, err error) {
+	client := p.client()
+	modelIdentifier = strings.TrimPrefix(modelIdentifier, p.ID()+provider.ProviderModelSeparator)
+
+	apiResponse, err := client.CreateChatCompletion(
```
Comment (Munsio) [inline provider/openai-api/openai.go]: In the previous commit you created `QueryOpenAIAPIModel` but are not using it in this one.

**Review by Munsio — CHANGES_REQUESTED** [review]

File: provider/openai-api/query.go (hunk 1 above)
Comment (bauersimon) [reply provider/openai-api/query.go:12]: Was thinking about it as well... but the openAI provider i.e. has a models slice an neither Ollama nor Openrouter need that. So... we can embed but it feels a bit mushy.

File: provider/openai-api/openai.go (hunk 3 above)
Comment (bauersimon) [reply provider/openai-api/openai.go]: I think you reviewed bottom-up. In GitHub the commits are chronologically top-down. Opposite of GitLab. :smiling_face_with_tear:

File: provider/openai-api/query.go (hunk 1 above)
Comment (Munsio) [reply provider/openai-api/query.go:12]: Hmm yeah I know what you mean. Alright then, ignore that one.

File: provider/openai-api/openai.go (hunk 3 above)
Comment (Munsio) [reply provider/openai-api/openai.go]: :facepalm:

**Pull request merged**